### ISSUE TITLE
Reporting intermediate solutions

### ISSUE THREAD

**Issue by IgnaceBleukx** [CPMpy/cpmpy#558]
For several use-cases where you solve hard optimization problems, it might be useful to get the intermediate solutions back from the solver.
I know Gurobi and OR-Tools support this feature through callbacks, but it requires a user to write its own wrapper for each solver he wants to use.

I think it would be nice if we can handle this in a more unified way.
If you agree, let's use this issue on how exactly this feature should be implemented.

My take:
Something similar to the `display` parameter in `solveAll` would be nice I think. 
Here, we can either pass a (list of) expressions whose value is printed after each solution found, or a "real" callback function that gets called on every solution. A user can use the callback himself to track time until this solution or similar stats.

**Comment by IgnaceBleukx** [comment]
Yes, that's indeed more or less what I envisioned : ) I'll make a branch and port some of the things you did for the competition onto that one.

**Comment by hbierlee** [comment]
In minizinc-python, you can actually iterate over the solutions, see https://python.minizinc.dev/en/latest/advanced_usage.html#concurrent-solutions, which might also be a nice design instead of callbacks:

```py
async for result in instance.solutions(all_solutions=True):
   ...
```
